### PR TITLE
Destroy models even when exceptions happen (adds 3 keep_model flags)

### PR DESCRIPTION
**Pull request by auria**
Models are preserved in two scenarios: 1) any model where tests fail (or any other environment-related problem is raised, except signals), 2) tests on a model successfully pass, but keep_model is defined and the model is the last one to be created.

This behavior can cause resources exhaustion on test environments, and logs may be enough to understand why tests failed. This change will try to remove models (if --keep-model is not defined) when an exception occurs, right after sharing a juju_status report.

Note: Models that are not the last one to be created will also be removed, no matter --keep-model is used (because only the last model is meant to be affected by --keep-model)

### CODE REVIEW

**Comment by ajkavanagh** [comment]
> When keep_model is not defined, models are only destroyed if tests were successful. Any errors will not remove the models, causing resources exhaustion on test environments. This change will try to remove models (if --keep-model is not defined) when an exception occurs, right after sharing a juju_status report.

This will be a good patch; it's probably been needed for a long time.  Thanks for looking in to this.

**Comment by auria** [comment]
Hey @ajkavanagh, I have changed the summary and description of the patch, because it assumed keep_model == --keep-model, but keep_model is in reality preserve_model, which is calculated from `keep_model and last_test == env_deployment.name`.

I have left this change as `Draft` because this part could be improved to effectively keep any failed model when `--keep-model` is shared (no matter it's not the last model created).

This patch could possible affect: https://github.com/openstack-charmers/zaza/issues/350, https://github.com/openstack-charmers/zaza/issues/284, https://github.com/openstack-charmers/zaza/issues/214.

Let me know what you think, and I can follow-up next week. Thank you for the quick reply.

**Comment by auria** [comment]
This has been tested using `TEST_DEPLOY_TIMEOUT=60` envvar and modifying `tox.ini`:

```
[testenv:func]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-model

[testenv:func-keep-all]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-all-models

[testenv:func-keep-faulty]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-faulty-model


[testenv:func-ambiguous1]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-model --keep-all-models

[testenv:func-ambiguous2]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-all-models --keep-faulty-model

[testenv:func-ambiguous3]
basepython = python3
deps = -r{toxinidir}/requirements.txt
commands =
    {envdir}/bin/python3 setup.py install
    functest-run-suite --keep-faulty-model --keep-model
```

**Review by ChrisMacNaughton — APPROVED** [review]
LGTM, would also like an additional pair of eyes :)

**Review by gnuoy — APPROVED** [review]
LGTM to me too

File: zaza/charm_lifecycle/func_test_runner.py [hunk 1]
```diff
@@ -81,14 +91,14 @@ def failure_report(model_aliases, show_juju_status=False):
                     default_flow_style=False))
 
 
-def run_env_deployment(env_deployment, keep_model=False, force=False,
+def run_env_deployment(env_deployment, keep_model=DESTROY_MODEL, force=False,
```
Comment (freyes) [inline zaza/charm_lifecycle/func_test_runner.py:94]: The argument `keep_model` changing from a boolean to a 'enum' makes it a bit counterintuitive (to me), the argument maybe should be renamed to `strategy` to make the signature of the function more obvious (Just a nitpick, feel free to disregard you disagree)

**Pull request merged**